upb/json/parser.rl

```diff
@@ -947,6 +973,134 @@ static bool start_stringval(upb_json_parser *p) {
 static bool end_stringval_nontop(upb_json_parser *p) {
   bool ok = true;
 
+  if (is_timestamp_object(p)) {
+    struct tm tm;
+    size_t len;
+    size_t i;
+    const char *buf = accumulate_getptr(p, &len);
+    const char *seconds_membername = "seconds";
+    const char *nanos_membername = "nanos";
+    char seconds_buf[20];
+    int64_t seconds = 0;
+    int32_t nanos = 0;
+
+    if (len < 19) {
+      upb_status_seterrf(&p->status, "error parsing timestamp");
+      upb_env_reporterror(p->env, &p->status);
+      multipart_end(p);
+      return false;
+    }
+
+    /* Parse seconds */
+    memcpy(seconds_buf, buf, 19);
+    seconds_buf[19] = 0;
+    if (strptime(seconds_buf, "%FT%H:%M:%S", &tm) == NULL) {
+      upb_status_seterrf(&p->status, "error parsing timestamp: %s",
+                         seconds_buf);
+      upb_env_reporterror(p->env, &p->status);
+      multipart_end(p);
+      return false;
+    }
+
+    /* Parse nanos */
+    if (len > 19 && buf[19] == '.') {
+      char *end;
+      char *nanos_buf;
+      char *nanos_bufend;
+      double val = 0.0;
+
+      i = 20;
+      while (i < len && buf[i] != 'Z' &&
+             buf[i] != '+' && buf[i] != '-') {
+        i++;
+      }
+
+      /* i - 19 is length of fractional part without starting '0'.
+       * Additional 2 for starting '0' and ending 0. */
+      nanos_buf = upb_gmalloc(i - 17);
+      nanos_buf[0] = '0';
+      memcpy(nanos_buf + 1, buf + 19, i - 19);
+      nanos_buf[i - 17] = 0;
+      nanos_bufend = nanos_buf + i - 18;
+
+      val = strtod(nanos_buf, &end);
+      if (errno == ERANGE || end != nanos_bufend) {
+        upb_status_seterrf(&p->status, "error parsing timestamp nanos: %s",
+                           nanos_buf);
+        upb_env_reporterror(p->env, &p->status);
+        upb_gfree(nanos_buf);
+        multipart_end(p);
+        return false;
+      }
+
+      nanos = val * 1000000000;
+      upb_gfree(nanos_buf);
+    } else {
+      i = 19;
+    }
+
+    /* Parse timezone */
+    if ((i == len - 1 && buf[i] == 'Z')) {
```

Comment:
This is a lot of manual parsing. We should be using Ragel for this if we can.

Is it possible that our "value" machine can enter a special "timestamp" machine if we decide this is a timestamp field? Then we could specify the format of a timestamp in Ragel, and have functions that get called for the different parts of a timestamp that set the appropriate values.